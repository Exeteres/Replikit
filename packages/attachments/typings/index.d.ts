export * from "./entities/attachment";

export * from "./extensions/storage";
