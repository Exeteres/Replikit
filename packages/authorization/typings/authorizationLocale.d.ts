export interface AuthorizationLocale {
    accessDenied: string;
}
