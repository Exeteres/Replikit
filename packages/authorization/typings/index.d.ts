export * from "./authorizarionLocale";
export * from "./extensions/i18n";
