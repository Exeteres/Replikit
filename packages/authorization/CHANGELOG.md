# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.14.0](https://github.com/Exeteres/Replikit/compare/v0.13.0...v0.14.0) (2020-08-05)

**Note:** Version bump only for package @replikit/authorization





# [0.13.0](https://github.com/Exeteres/Replikit/compare/v0.12.1...v0.13.0) (2020-08-04)


### Bug Fixes

* **authorization:** Fix some bugs in authorization system ([5d0f0d3](https://github.com/Exeteres/Replikit/commit/5d0f0d339c41e3982b9f165685101fd6a3a6998c))





# [0.12.0](https://github.com/Exeteres/Replikit/compare/v0.11.2...v0.12.0) (2020-07-30)


### Bug Fixes

* **authorization:** Fix extensions to use new core types ([c756c73](https://github.com/Exeteres/Replikit/commit/c756c73edb0072760789846bf8e43894b25d4f27))


### Features

* **authorization:** Redesign authorization system ([6dd23d1](https://github.com/Exeteres/Replikit/commit/6dd23d1b0bf830bb7dcf8007987d671bd2814c92))





## [0.11.2](https://github.com/Exeteres/Replikit/compare/v0.11.1...v0.11.2) (2020-07-24)

**Note:** Version bump only for package @replikit/authorization





## [0.11.1](https://github.com/Exeteres/Replikit/compare/v0.11.0...v0.11.1) (2020-07-24)

**Note:** Version bump only for package @replikit/authorization






# [0.11.0](https://github.com/Exeteres/Replikit/compare/v0.10.1...v0.11.0) (2020-07-24)


### Features

* **i18n:** Replace way to use i18n ([df6abcd](https://github.com/Exeteres/Replikit/commit/df6abcd791644bc2255107d98497a7e492fb07ca))





## [0.10.1](https://github.com/Exeteres/Replikit/compare/v0.10.0...v0.10.1) (2020-06-30)

**Note:** Version bump only for package @replikit/authorization






# [0.10.0](https://github.com/Exeteres/Replikit/compare/v0.9.0...v0.10.0) (2020-06-29)

**Note:** Version bump only for package @replikit/authorization





# [0.9.0](https://github.com/Exeteres/Replikit/compare/v0.8.0...v0.9.0) (2020-06-28)

**Note:** Version bump only for package @replikit/authorization





# [0.8.0](https://github.com/Exeteres/Replikit/compare/v0.7.2...v0.8.0) (2020-06-27)


### Bug Fixes

* **authorization:** Fix commands extension ([aa9b255](https://github.com/Exeteres/Replikit/commit/aa9b25592201318a9e97ab57f72fb9c11364d84e))





## [0.7.2](https://github.com/Exeteres/Replikit/compare/v0.7.1...v0.7.2) (2020-06-23)

**Note:** Version bump only for package @replikit/authorization





## [0.7.1](https://github.com/Exeteres/Replikit/compare/v0.7.0...v0.7.1) (2020-06-23)


### Bug Fixes

* **storage:** Fix invalid typings ([fef2cf0](https://github.com/Exeteres/Replikit/commit/fef2cf0be233d99802829c65527a9934a2371ab0))





# [0.7.0](https://github.com/Exeteres/Replikit/compare/v0.6.0...v0.7.0) (2020-06-23)


### Bug Fixes

* **authorization:** Fix extension typings and accessDenied message formatting ([fc28542](https://github.com/Exeteres/Replikit/commit/fc28542dcb84f82d20a4552ad31b6b89fb0b9ec7))





# [0.6.0](https://github.com/Exeteres/Replikit/compare/v0.5.3...v0.6.0) (2020-06-21)


### Bug Fixes

* Fix typings ([315e2b1](https://github.com/Exeteres/Replikit/commit/315e2b1fedab922dd3e63851f33de64856a8d863))





## [0.5.3](https://github.com/Exeteres/Replikit/compare/v0.5.2...v0.5.3) (2020-06-20)

**Note:** Version bump only for package @replikit/authorization





## [0.5.2](https://github.com/Exeteres/Replikit/compare/v0.5.1...v0.5.2) (2020-06-20)

**Note:** Version bump only for package @replikit/authorization





# 0.5.0 (2020-06-20)

**Note:** Version bump only for package @replikit/authorization






# [0.4.0](https://github.com/Exeteres/Replikit/compare/v0.3.0...v0.4.0) (2020-06-19)

**Note:** Version bump only for package @replikit/authorization





# [0.3.0](https://github.com/Exeteres/Replikit/compare/v0.2.0...v0.3.0) (2020-06-14)

**Note:** Version bump only for package @replikit/authorization





# [0.2.0](https://github.com/Exeteres/Replikit/compare/v0.1.0...v0.2.0) (2020-06-13)

**Note:** Version bump only for package @replikit/authorization
