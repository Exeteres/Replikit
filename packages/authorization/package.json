{
    "name": "@replikit/authorization",
    "version": "0.5.3",
    "license": "MIT",
    "main": "dist/index.js",
    "typings": "dist/index.d.ts",
    "files": [
        "dist",
        "typings"
    ],
    "scripts": {
        "build": "ttsc -p tsconfig.build.json"
    },
    "dependencies": {
        "@replikit/commands": "^0.5.3",
        "@replikit/permissions": "^0.5.3"
    }
}
