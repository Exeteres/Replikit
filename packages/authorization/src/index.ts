/// <reference path="../typings/index.d.ts" />

import "@replikit/permissions";

export * from "./commandBuilderExtension";

import "./locales/en";
import "./locales/ru";
