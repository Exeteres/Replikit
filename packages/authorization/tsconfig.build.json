{
    "extends": "../../tsconfig.build.json",
    "compilerOptions": {
        "outDir": "dist",
        "paths": {
            "@replikit/authorization": ["authorization/src"]
        }
    },
    "include": ["src"]
}
