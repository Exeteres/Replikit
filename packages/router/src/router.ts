import { Event, EventName } from "@replikit/core/typings";
import {
    HandlerChain,
    ContextFactoryStorage,
    contextFactories
} from "@replikit/router";
import { ContextMap } from "@replikit/router/typings";

type RouterChain = HandlerChain<ContextMap[keyof ContextMap]>;

export class Router {
    private readonly chainMap = new Map<EventName, RouterChain>();

    constructor(private readonly contextFactories: ContextFactoryStorage) {}

    of<T extends EventName>(type: T): HandlerChain<ContextMap[T]> {
        let chain = this.chainMap.get(type);
        if (!chain) {
            chain = new HandlerChain();
            this.chainMap.set(type, chain);
        }
        return (chain as unknown) as HandlerChain<ContextMap[T]>;
    }

    async process(event: Event): Promise<void> {
        const chain = this.chainMap.get(event.type);
        if (!chain) {
            return;
        }
        const context = await this.contextFactories.createContext(event);
        return chain.process(context);
    }
}

export const router = new Router(contextFactories);
