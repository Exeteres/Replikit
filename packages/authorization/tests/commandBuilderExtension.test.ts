import "@replikit/authorization";
import {
    DatabaseTestManager,
    TestManagerSuite,
    createTestManager
} from "@replikit/test-utils";
import { MessageContext } from "@replikit/router";
import {
    AccountContextExtension,
    User,
    Account,
    Member,
    Channel
} from "@replikit/storage";
import { UserPermissionName } from "@replikit/permissions/typings";
import { fromText } from "@replikit/messages";

let dbTestManager: DatabaseTestManager;

beforeEach(() => {
    dbTestManager = new DatabaseTestManager();
    return dbTestManager.connect();
});

afterEach(() => {
    return dbTestManager.close();
});

function createExtensionTestManager(): TestManagerSuite {
    const suite = createTestManager();
    const manager = dbTestManager;
    suite.testManager.contextFactories.register("message:received", event => {
        const context = new MessageContext(event);
        ((context as unknown) as AccountContextExtension)._connection =
            manager.connection;
        return context;
    });
    return suite;
}

async function insertUser(
    id: number,
    permissions: UserPermissionName[]
): Promise<void> {
    const collection = dbTestManager.connection.getCollection(User);
    await collection.insertOne(({
        permissions,
        roles: [],
        accounts: [{ controller: "test", localId: id } as Account]
    } as unknown) as User);
}

async function insertMember(
    accountId: number,
    permissions: UserPermissionName[],
    channelId = 0,
    controller = "test"
): Promise<void> {
    const collection = dbTestManager.connection.getCollection(Member);
    await collection.insertOne({
        _id: { controller, channelId, accountId },
        permissions,
        roles: []
    });
}

async function insertChannel(
    channelId: number,
    controller = "test"
): Promise<void> {
    const repository = dbTestManager.connection.getRepository(Channel);
    const channel = repository.create({ controller, localId: channelId });
    await channel.save();
}

describe("CommandBuilderExtension", () => {
    it("should authorize a user permission", async () => {
        const { testManager, command } = createExtensionTestManager();

        await insertUser(1, ["test1"]);
        await insertUser(2, []);

        command("test")
            .authorizeUser("test1")
            .handler(() => fromText("Access allowed"))
            .register();

        await testManager.processCommand("/test", 1);
        await testManager.processCommand("/test", 2);
        await testManager.processCommand("/test", 3);

        expect.assertions(3);
    });

    it("should authorize a member permission", async () => {
        const { testManager, command } = createExtensionTestManager();

        await insertMember(1, ["test1"]);
        await insertMember(2, []);

        command("test")
            .authorizeMember("test1")
            .handler(() => fromText("Access allowed"))
            .register();

        await testManager.processCommand("/test", 1);
        await testManager.processCommand("/test", 2);
        await testManager.processCommand("/test", 3);

        expect.assertions(3);
    });

    it("should authorize a member permission in another channel of the same controller", async () => {
        const { testManager, command } = createExtensionTestManager();

        await insertMember(1, ["test1"], 2);
        await insertMember(2, [], 2);

        await insertChannel(2);

        command("test")
            .channel()
            .authorizeMember("test1")
            .handler(() => fromText("Access allowed"))
            .register();

        await testManager.processCommand("/test 1", 1);
        await testManager.processCommand("/test 1", 2);
        await testManager.processCommand("/test 1", 3);

        expect.assertions(3);
    });

    it("should authorize a member permission in the same channel", async () => {
        const { testManager, command } = createExtensionTestManager();

        await insertMember(1, ["test1"], 0);
        await insertMember(2, [], 0);

        await insertChannel(0);

        command("test")
            .channel()
            .authorizeMember("test1")
            .handler(() => fromText("Access allowed"))
            .register();

        await testManager.processCommand("/test", 1);
        await testManager.processCommand("/test", 2);
        await testManager.processCommand("/test", 3);

        expect.assertions(3);
    });

    it("should authorize a member permission in another channel of another controller", async () => {
        const { testManager, command } = createExtensionTestManager();

        const users = dbTestManager.connection.getCollection(User);
        await users.insertOne(({
            permissions: [],
            roles: [],
            accounts: [
                { controller: "test", localId: 1 },
                { controller: "another", localId: 2 }
            ]
        } as unknown) as User);

        await insertChannel(2, "another");
        await insertMember(2, ["test1"], 2, "another");

        command("test")
            .channel()
            .authorizeMember("test1")
            .handler(() => fromText("Access allowed"))
            .register();

        await testManager.processCommand("/test 1", 1);
        await testManager.processCommand("/test 1", 2);

        expect.assertions(2);
    });
});
