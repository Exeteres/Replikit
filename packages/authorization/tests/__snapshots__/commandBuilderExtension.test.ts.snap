// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandBuilderExtension should authorize a member permission 1`] = `"Access allowed"`;

exports[`CommandBuilderExtension should authorize a member permission 2`] = `"Access denied"`;

exports[`CommandBuilderExtension should authorize a member permission 3`] = `"Access denied"`;

exports[`CommandBuilderExtension should authorize a member permission in another channel of another controller 1`] = `"Access allowed"`;

exports[`CommandBuilderExtension should authorize a member permission in another channel of another controller 2`] = `"Access denied"`;

exports[`CommandBuilderExtension should authorize a member permission in another channel of the same controller 1`] = `"Access allowed"`;

exports[`CommandBuilderExtension should authorize a member permission in another channel of the same controller 2`] = `"Access denied"`;

exports[`CommandBuilderExtension should authorize a member permission in another channel of the same controller 3`] = `"Access denied"`;

exports[`CommandBuilderExtension should authorize a member permission in the same channel 1`] = `"Access allowed"`;

exports[`CommandBuilderExtension should authorize a member permission in the same channel 2`] = `"Access denied"`;

exports[`CommandBuilderExtension should authorize a member permission in the same channel 3`] = `"Access denied"`;

exports[`CommandBuilderExtension should authorize a user permission 1`] = `"Access allowed"`;

exports[`CommandBuilderExtension should authorize a user permission 2`] = `"Access denied"`;

exports[`CommandBuilderExtension should authorize a user permission 3`] = `"Access denied"`;
